### PR TITLE
Added support to 'upload_port' option on platformio.ini and picoprobe upload through upload_protocol = picoprobe

### PR DESCRIPTION
**Pull request by Waaflee**
Added support to 'upload_port' on option on platformio.ini and picoprobe upload through upload_protocol = picoprobe

now upload_port setting works, if device is not found it graciously falls back to automatic discovery.

now upload_protocol = picoprobe works, it flashes the device through openocd protocol (picoprobe)

both features tested on Archlinux and Windows 10/11 successfully.

### CODE REVIEW

File: builder/frameworks/uf2conv.py [hunk 1]
```diff
@@ -318,30 +346,77 @@ def dev_uploader(target, source, env):
...
+        # Gets Picoprobe configuration
+        picoprobe = env.BoardConfig().__dict__[
+            "_manifest"]["debug"]["tools"]["picoprobe"]
+        
+        # path to picoprobe binary
+        path = join(env.PioPlatform().get_package_dir(
+            'tool-pico-openocd'), picoprobe["server"]["executable"])
+        
+        # extra arguments to configure openocd
+        arguments = ' '.join(picoprobe["server"]["arguments"]).replace("$PACKAGE_DIR", env.PioPlatform().get_package_dir(
```
Comment (flaviut) [inline builder/frameworks/uf2conv.py:382]: best practice is to keep each command parameter in its own string in a list.

```suggestion
        ocd_package_dir = env.PioPlatform().get_package_dir('tool-pico-openocd')
        arguments = [arg.replace("$PACKAGE_DIR", ocd_package_dir) for arg in picoprobe["server"]["arguments"]]
```

Joining everything into one big string will, at best, cause issues when someone uses a space in a filename. At worst, it'll allow for execution of arbitrary programs and is a security risk.

File: builder/frameworks/uf2conv.py [hunk 2]
```diff
@@ -318,30 +346,77 @@ def dev_uploader(target, source, env):
...
+        # openocd run command, upload .elf binary
+        order = f"-c 'program {elf_name} verify reset exit'"
+       
+        command = ' '.join([path, arguments, order])
+        if sys.platform == "win32":
+            command = command.replace("\\", "/").replace("'", "\"")
+
+        # TODO: this could be improved (less verbose, better error handling, etc)
+        print("Flashing through picoprobe...")
+        os.system(command)
```
Comment (flaviut) [inline builder/frameworks/uf2conv.py:394]:
```suggestion
        subprocess.run(arguments)

```

File: builder/frameworks/uf2conv.py [hunk 3]
```diff
@@ -318,30 +346,77 @@ def dev_uploader(target, source, env):
...
+    print("  Converting to UF2 ( 0x%x )" % (appstartaddr))
+    with open(bin_name, mode='rb') as f:
+        inpbuf = f.read()
     outbuf = convert_to_uf2(inpbuf)
     time.sleep(.1)
-    write_file(uf2_name, outbuf) # write uf2 to build folder
+    write_file(uf2_name, outbuf)  # write uf2 to build folder
+
+    # Uploads via picoprobe protocol
+    if env.GetProjectOption("upload_protocol") == "picoprobe":
```
Comment (flaviut) [inline builder/frameworks/uf2conv.py:370]: This functionality doesn't make use of the u2f system at all--seems to me like it'd be better if it was a separate function in a separate file and inserted into the config here: https://github.com/Wiz-IO/wizio-pico/blob/main/builder/frameworks/common.py#L180

**Review by flaviut — COMMENTED** [review]
Very cool work--and greatly appreciated! I'll be using this in my personal fork.

I noticed a few things that could be improved, but I'm not the maintainer of this package, and I don't know if he has different opinions than me :)

**Comment by casundra** [comment]
Just chiming in to say that I would love to see this merged into the main branch!  I've been using Wizio so that I can work with the SDK directly, and it's great.  I would love to be able to do everything with one USB port, instead of having to connect both the target Pico and PicoProbe Pico.

**Comment by flaviut** [comment]
Here's my current config if you're interested:

```
[env:rp2040]
platform = https://github.com/flaviut/wizio-pico#0f237725ea45cd1501f9fddd3be4e4c9de8a8b4c
board = raspberry-pi-pico
framework = baremetal
# disable unknown pragma warnings, since those are just from our code quality tools
build_flags =
    -std=c++17 -Wno-unknown-pragmas
    -D TARGET_RP20xx
    -D LIB_PICO_STDIO_USB
     # enable hardware acceleration
    -D PICO_DIVIDER_IN_RAM -D PICO_INT64_OPS_IN_RAM -D PICO_MEM_IN_RAM
    -D PICO_DIVIDER_HARDWARE -D PICO_INT64_OPS_PICO -D PICO_MEM_OPS_PICO
    # -D PICO_BIT_OPS_PICO disabled due to bug: https://github.com/raspberrypi/pico-sdk/issues/765
debug_tool = picoprobe
upload_protocol = picoprobe
```

It also uses a more modern compiler & some other small tweaks I can't remember.

**Pull request closed without merging**